packages/nouns-contracts/contracts/Rewards.sol

```diff
@@ -0,0 +1,280 @@
+            t.numEligibleVotes += votesInProposal;
+        }
+
+        //// Check that distribution is allowed:
+        //// 1. At least one eligible proposal.
+        //// 2. One of the two conditions must be true:
+        //// 2.a. Number of eligible proposals is at least `numProposalsEnoughForReward`.
+        //// 2.b. At least `minimumRewardPeriod` seconds have passed since the last update.
+
+        require(numEligibleProposals > 0, 'at least one eligible proposal');
+        if (numEligibleProposals < params.numProposalsEnoughForReward) {
+            require(
+                lastProposal.creationTimestamp > lastProposalRewardsUpdate + params.minimumRewardPeriod,
+                'not enough time passed'
+            );
+        }
+        lastProposalRewardsUpdate = lastProposal.creationTimestamp;
+
+        // Calculate the reward per proposal and per vote
+        t.rewardPerProposal = proposalRewardForPeriod / numEligibleProposals;
+        t.rewardPerVote = votingRewardForPeriod / t.numEligibleVotes;
+
+        //// Second loop over the proposals:
+        //// 1. Skip proposals that were deleted for non eligibility.
+        //// 2. Reward proposal's clientId.
+        //// 3. Reward the clientIds that faciliated voting.
+        //// 4. Make sure all voting clientIds were included.
+
+        for (uint256 i; i < proposals.length; ++i) {
+            // skip non eligible deleted proposals
+            if (proposals[i].endBlock == 0) continue;
+
+            uint32 clientId = proposals[i].clientId;
+            if (clientId != 0) {
+                clientBalances[clientId] += t.rewardPerProposal;
+            }
+
+            uint256 votesInProposal;
+            NounsDAOV3Types.ClientVoteData[] memory voteData = proposals[i].voteData;
+            uint256 votes;
+            for (uint256 j; j < votingClientIds.length; ++j) {
+                clientId = votingClientIds[j];
+                votes = voteData[j].votes;
+                if (clientId != 0) {
+                    clientBalances[clientId] += votes * t.rewardPerVote;
+                }
+                votesInProposal += votes;
+            }
+            require(
+                votesInProposal == proposals[i].forVotes + proposals[i].againstVotes + proposals[i].abstainVotes,
+                'not all votes accounted'
+            );
+        }
+    }
+
+    function withdrawClientBalance(uint32 clientId, uint256 amount, address to) public {
+        require(ownerOf(clientId) == msg.sender, 'must be client NFT owner');
+
+        clientBalances[clientId] -= amount;
+
+        ethToken.transfer(to, amount);
+    }
+
+    struct ProposalRewardsParams {
+        uint32 lastProposalId;
+        uint256 expectedNumEligibleProposals;
+        uint256 expectedNumEligibleVotes;
+        uint256 firstNounId;
+        uint256 lastNounId;
+        uint32[] votingClientIds;
+    }
+
+    function getAuctionRevenueBetweenNouns(
```

Comment:
wdyt of adding such a view function to AH? something like `getSettlements(firstNounId, endTimestamp)`?

Reply:
@eladmallel thinking about how it would work:
it could loop through all settled auction ids from `firstNounId` upwards, and stop when it reaches a timestamp > endTimestamp.
a bit annoying that it doesn't know the final array size in advance, but I'll give it a go. would be nice to remove the `lastNounId` parameter!

Reply:
implemented!